### ISSUE TITLE
Possible mitogen bug with low address space in /etc/security/limits.conf

### ISSUE THREAD

**Issue by bergsjoh** [dw/mitogen#394]
As discussed abit on IRC, Crashed when running anything (tested -m ping and -m command -a free) in ansible with mitogen transport when address space is limited to below 1248000kb in /etc/security/limits.conf

The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 138, in run
    res = self._execute()
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 576, in _execute
    result = self._handler.run(task_vars=variables)
  File "/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/mixins.py", line 115, in run
    return super(ActionModuleMixin, self).run(tmp, task_vars)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/normal.py", line 46, in run
    result = merge_hash(result, self._execute_module(task_vars=task_vars, wrap_async=wrap_async))
  File "/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/mixins.py", line 345, in _execute_module
    timeout_secs=self.get_task_timeout_secs(),
  File "/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/planner.py", line 479, in invoke
    kwargs=planner.get_kwargs(),
  File "/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/connection.py", line 594, in call
    return self.call_async(func, *args, **kwargs).get().unpickle()
  File "/home/bergsjoh/ansible/strategy_plugins/mitogen/core.py", line 487, in unpickle
    raise obj
**CallError: thread.error: can't start new thread**
  File "<stdin>", line 2049, in _dispatch_calls
  File "<stdin>", line 2041, in _dispatch_one
  File "master:/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/target.py", line 272, in run_module
    return impl.run()
  File "master:/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/runner.py", line 225, in run
    self.setup()
  File "master:/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/runner.py", line 552, in setup
    super(NewStyleRunner, self).setup()
  File "master:/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/runner.py", line 349, in setup
    self._setup_program()
  File "master:/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/runner.py", line 584, in _setup_program
    path=self.path,
  File "master:/home/bergsjoh/ansible/strategy_plugins/ansible_mitogen/target.py", line 85, in get_small_file
    pool = mitogen.service.get_or_create_pool(router=context.router)
  File "master:/home/bergsjoh/ansible/strategy_plugins/mitogen/service.py", line 67, in get_or_create_pool
    _pool = Pool(router, [], size=size or DEFAULT_POOL_SIZE)
  File "master:/home/bergsjoh/ansible/strategy_plugins/mitogen/service.py", line 453, in __init__
    thread.start()
  File "/usr/lib64/python2.6/threading.py", line 474, in start
    _start_new_thread(self.__bootstrap, ())

Workaround is ofcourse to increase address space, but it was working fine when running standard Ansible transport so reported here.

**Comment by dw** [comment]
Reproduction:

````
/usr/bin/time -v bash -c "( ulimit -v 1248000 ; ANSIBLE_STRATEGY=mitogen ansible -vvvm setup k3; )"
````

**Comment by dw** [comment]
Per IRC, limiting address space is a bad idea, but this still warrants investigation as to why it's hitting the 1GB limit.

It looks like the problem may be glibc's memory allocator, which I suspect is constructing a new 64MB mapping for each thread in the connection multiplexer. 64MB * 16 = 1GB of almost entirely unallocated, but nonetheless mapped memory.

Still digging to find out who owns those 64MB segments, but 90% sure it's malloc

**Comment by dw** [comment]
On Linux, a nasty workaround is to ensure MALLOC_ARENA_MAX=1 environment variable is set prior to starting Ansible. This may hinder performance somewhat, but it avoids relying on large chunks of address space existing

**Comment by dw** [comment]
Hi Johan,

I'm going to close this off, as it doesn't seem like there is much to be done here. Much of a modern system relies on virtual memory for all kinds of trickery, such as lazily-allocated malloc arenas, and restricting virtual memory usage simply breaks far too much stuff :)

Regarding the original use case we discussed, even with a limit as low as ~1GB, this still permits a hostile process to allocate >300k VMA structures in the kernel, which is entirely not accounted for using the ulimit mechanism.

It's definitely worth exploring either the cgroups memory controller (if available) to implement your desired restriction. It turns out ( https://linux.die.net/man/2/setrlimit ) RSS restrictions are entirely ignored on modern Linux.

Thanks again for reporting this -- I got to learn something new about glibc in the process :)
